spark/v4.0/spark/src/main/java/org/apache/iceberg/spark/actions/SparkParquetFileMergeRunner.java

```diff
@@ -0,0 +1,460 @@
+  @Override
+  public String description() {
+    return "PARQUET-MERGE";
+  }
+
+  @Override
+  protected void doRewrite(String groupId, RewriteFileGroup group) {
+    // Early validation: check if requirements are met
+    if (!canMerge(group)) {
+      LOG.info(
+          "Row-group merge requirements not met for group {}. Using standard Spark rewrite.",
+          groupId);
+      super.doRewrite(groupId, group);
+      return;
+    }
+
+    // Requirements met - attempt row-group level merge
+    try {
+      LOG.info(
+          "Merging {} Parquet files using row-group level merge (group: {})",
+          group.rewrittenFiles().size(),
+          groupId);
+      mergeParquetFilesDistributed(groupId, group);
+    } catch (Exception e) {
+      LOG.warn(
+          "Row-group merge failed for group {}, falling back to standard Spark rewrite: {}",
+          groupId,
+          e.getMessage(),
+          e);
+      // Fallback to standard rewrite
+      super.doRewrite(groupId, group);
+    }
+  }
+
+  /**
+   * Validates if the file group can use row-group level merging.
+   *
+   * <p>Requirements checked here:
+   *
+   * <ul>
+   *   <li>Table must not have a sort order (no sorting or z-ordering)
+   *   <li>Files must not have delete files or delete vectors
+   * </ul>
+   *
+   * <p>Additional requirements checked by ParquetFileMerger.canMerge:
+   *
+   * <ul>
+   *   <li>All files must be valid Parquet format
+   *   <li>Files must have compatible schemas
+   *   <li>Files must not be encrypted
+   *   <li>If a physical _row_id column exists, all values must be non-null
+   * </ul>
+   *
+   * @param group the file group to check
+   * @return true if row-group merging can be used, false otherwise
+   */
+  @VisibleForTesting
+  boolean canMerge(RewriteFileGroup group) {
+    // Check if table has a sort order - row-group merge cannot preserve sort order
+    if (table().sortOrder().isSorted()) {
+      LOG.debug(
+          "Cannot use row-group merge: table has a sort order ({}). "
+              + "Row-group merging would not preserve the sort order.",
+          table().sortOrder());
+      return false;
+    }
+
+    // Check for delete files - row-group merge cannot apply deletes
+    boolean hasDeletes = group.fileScanTasks().stream().anyMatch(task -> !task.deletes().isEmpty());
+
+    if (hasDeletes) {
+      LOG.debug(
+          "Cannot use row-group merge: files have delete files or delete vectors. "
+              + "Row-group merging cannot apply deletes.");
+      return false;
+    }
+
+    // Validate schema compatibility and other Parquet-specific requirements
+    try {
+      List<InputFile> inputFiles =
+          group.rewrittenFiles().stream()
+              .map(f -> table().io().newInputFile(f.path().toString()))
+              .collect(Collectors.toList());
+
+      // Validate files can be merged
+      boolean canMerge = ParquetFileMerger.canMerge(inputFiles);
+
+      if (!canMerge) {
+        LOG.warn(
```

Comment:
This is info. Warn means something not expected happens

Reply:
ok